Jenkinsfile

```diff
@@ -43,458 +49,599 @@ pipeline {
-                      -Dsonar.github.oauth=${SORABOT_TOKEN}
-                  """
-                }
-                sh "cmake --build build --target coverage.info"
-                sh "python /usr/local/bin/lcov_cobertura.py build/reports/coverage.info -o build/reports/coverage.xml"
-                cobertura autoUpdateHealth: false, autoUpdateStability: false, coberturaReportFile: '**/build/reports/coverage.xml', conditionalCoverageTargets: '75, 50, 0', failUnhealthy: false, failUnstable: false, lineCoverageTargets: '75, 50, 0', maxNumberOfBuilds: 50, methodCoverageTargets: '75, 50, 0', onlyStable: false, zoomCoverageChart: false
+              if ( params.BUILD_TYPE == 'Debug' || params.Merge_PR ) {
+                def macDebugBuild = load ".jenkinsci/mac-debug-build.groovy"
+                def coverage = load ".jenkinsci/selected-branches-coverage.groovy"
+                macDebugBuild.doDebugBuild()
               }
-              if (GIT_LOCAL_BRANCH ==~ /(master|develop)/) {
-                releaseBuild = load ".jenkinsci/mac-release-build.groovy"
-                releaseBuild.doReleaseBuild()
+              else {
+                def macReleaseBuild = load ".jenkinsci/mac-release-build.groovy"
+                macReleaseBuild.doReleaseBuild()
               }
             }
           }
           post {
-            always {
+            success {
               script {
-                timeout(time: 600, unit: "SECONDS") {
-                  try {
-                    if (currentBuild.currentResult == "SUCCESS" && GIT_LOCAL_BRANCH ==~ /(master|develop)/) {
-                      def artifacts = load ".jenkinsci/artifacts.groovy"
-                      def commit = env.GIT_COMMIT
-                      filePaths = [ '\$(pwd)/build/*.tar.gz' ]
-                      artifacts.uploadArtifacts(filePaths, sprintf('/iroha/macos/%1$s-%2$s-%3$s', [GIT_LOCAL_BRANCH, sh(script: 'date "+%Y%m%d"', returnStdout: true).trim(), commit.substring(0,6)]))
-                    }
-                  }
-                  finally {
-                    cleanWs()
-                    sh """
-                      pg_ctl -D /var/jenkins/${GIT_COMMIT}-${BUILD_NUMBER}/ stop && \
-                      rm -rf /var/jenkins/${GIT_COMMIT}-${BUILD_NUMBER}/
-                    """
-                  }
+                if (params.BUILD_TYPE == 'Release') {
+                  def post = load ".jenkinsci/linux-post-step.groovy"
+                  post.macPostStep()
                 }
               }
             }
           }
         }
       }
     }
-    stage('Build Release') {
+    stage('Pre-Coverage') {
+      when {
+        anyOf {
+          expression { params.Coverage }  // by request
+          allOf {
+            expression { return env.CHANGE_ID }
+            expression { return !env.GIT_PREVIOUS_COMMIT }
+          }
+          allOf {
+            expression { return env.CHANGE_ID != null }
```

Comment:
After discussion with @tyvision we decided to move these complex condition to separate variables as `SOME_VAR = [[ env.CHANGE_ID != null && env.CHANGE_TARGET ==~ /(master|develop|trunk)/ ]]`, and then use them here. These variables will have documentation where they are defined, and the logic will not be duplicated in different stages.